File: BackEnd/TypeScript/nestshop/src/badge/badge.controller.ts
import { Body, Controller, Delete, Get, Param, ParseIntPipe, Patch, Post } from '@nestjs/common';
import { ApiBearerAuth, ApiOperation, ApiTags } from '@nestjs/swagger';
import { CurrentUser, JwtPayload } from '../common/decorators/current-user.decorator';
import { Public } from '../common/decorators/public.decorator';
import { Roles, UserRole } from '../common/decorators/roles.decorator';
import { BadgeService } from './badge.service';
import { CreateBadgeDto } from './dto/create-badge.dto';
import { GrantBadgeDto } from './dto/grant-badge.dto';
import { UpdateBadgeDto } from './dto/update-badge.dto';

@ApiTags('Badge')
@Controller()
export class BadgeController {
  constructor(private readonly badgeService: BadgeService) {}

  @Public()
  @Get('badges')
  @ApiOperation({ summary: '전체 배지 목록 조회' })
  getAllBadges() {
    return this.badgeService.getAllBadges();
  }

  @ApiBearerAuth()
  @Get('badges/me')
  @ApiOperation({ summary: '내 배지 목록 조회' })
  getMyBadges(@CurrentUser() user: JwtPayload) {
    return this.badgeService.getMyBadges(user.sub);
  }

  @Public()
  @Get('users/:id/badges')
  @ApiOperation({ summary: '특정 유저 배지 목록 조회' })
  getUserBadges(@Param('id', ParseIntPipe) userId: number) {
    return this.badgeService.getUserBadges(userId);
  }

  @ApiBearerAuth()
  @Roles(UserRole.ADMIN)
  @Post('admin/badges')
  @ApiOperation({ summary: '배지 생성 (Admin)' })
  create(@Body() dto: CreateBadgeDto) {
    return this.badgeService.create(dto);
  }

  @ApiBearerAuth()
  @Roles(UserRole.ADMIN)
  @Patch('admin/badges/:id')
  @ApiOperation({ summary: '배지 수정 (Admin)' })
  update(@Param('id', ParseIntPipe) id: number, @Body() dto: UpdateBadgeDto) {
    return this.badgeService.update(id, dto);
  }

  @ApiBearerAuth()
  @Roles(UserRole.ADMIN)
  @Delete('admin/badges/:id')
  @ApiOperation({ summary: '배지 삭제 (Admin)' })
  remove(@Param('id', ParseIntPipe) id: number) {
    return this.badgeService.remove(id);
  }

  @ApiBearerAuth()
  @Roles(UserRole.ADMIN)
  @Post('admin/badges/:id/grant')
  @ApiOperation({ summary: '배지 수동 부여 (Admin)' })
  grant(
    @Param('id', ParseIntPipe) badgeId: number,
    @Body() dto: GrantBadgeDto,
    @CurrentUser() user: JwtPayload,
  ) {
    return this.badgeService.grant(badgeId, dto, user.sub);
  }

  @ApiBearerAuth()
  @Roles(UserRole.ADMIN)
  @Delete('admin/badges/:id/revoke/:userId')
  @ApiOperation({ summary: '배지 회수 (Admin)' })
  revoke(@Param('id', ParseIntPipe) badgeId: number, @Param('userId', ParseIntPipe) userId: number) {
    return this.badgeService.revoke(badgeId, userId);
  }
}

File: BackEnd/Kotlin/kotlin-ktor-gradle-exposeddao-postgresql/src/main/kotlin/com/pbshop/kotlin/ktor/gradle/exposeddao/postgresql/badge/BadgeController.kt
package com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.badge

import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.pbActor
import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.respondStub
import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.security.PbRole
import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.security.requireAnyRole
import io.ktor.server.application.call
import io.ktor.server.request.receive
import io.ktor.server.routing.Route
import io.ktor.server.routing.delete
import io.ktor.server.routing.get
import io.ktor.server.routing.patch
import io.ktor.server.routing.post

class BadgeController(
    private val service: BadgeService,
) {
    fun Route.register() {
        get("/badges") {
            call.respondStub(service.listBadges())
        }
        get("/badges/me") {
            call.requireAnyRole(PbRole.USER)
            call.respondStub(service.myBadges(call.pbActor().userId ?: 0))
        }
        get("/users/{id}/badges") {
            call.respondStub(service.userBadges(call.parameters["id"]?.toIntOrNull() ?: 0))
        }
        post("/admin/badges") {
            call.requireAnyRole(PbRole.ADMIN)
            call.respondStub(service.create(call.receive()))
        }
        patch("/admin/badges/{id}") {
            call.requireAnyRole(PbRole.ADMIN)
            call.respondStub(service.update(call.parameters["id"]?.toIntOrNull() ?: 0, call.receive()))
        }
        delete("/admin/badges/{id}") {
            call.requireAnyRole(PbRole.ADMIN)
            call.respondStub(service.delete(call.parameters["id"]?.toIntOrNull() ?: 0))
        }
        post("/admin/badges/{id}/grant") {
            call.requireAnyRole(PbRole.ADMIN)
            call.respondStub(service.grant(call.parameters["id"]?.toIntOrNull() ?: 0, call.receive(), call.pbActor().userId))
        }
        delete("/admin/badges/{id}/revoke/{userId}") {
            call.requireAnyRole(PbRole.ADMIN)
            call.respondStub(
                service.revoke(
                    badgeId = call.parameters["id"]?.toIntOrNull() ?: 0,
                    userId = call.parameters["userId"]?.toIntOrNull() ?: 0,
                ),
            )
        }
    }
}
